### PR TITLE
Alt click refactor

### LINKED ISSUE #81242: Okay a whole lot of things lack proper sanity checks for alt-clicking
## Issue Summary

- Tape recorder (partially)
- Autolathe
- Pillow
- Ethereal Disco Ball
- Health Assembly
- Medical roller bed
- Toilet Bong
- Cigarette boxes
- Wrapping Paper
- Medical Mesh (partially0
- Cyborg organ storage
- Cyborg apparatuses in general
- Borg Hypospray
- Rapid Wiring Device
- Plumbing Device
- Electrolyzer (partially)
- Thermomachine (partially)
- Universal Scanner
- Glasses / Debug glasses
- Inspector Gadget fedora
- Animal mask
- Bandanas (partially)
- Neck ties
- Uniforms (partially)
- Hotel door (partially)
- Goliath Cloak (partially) 
- Clipboard
- Papercutter (partially)
- Bow (partially)
- Chem mass spectrometer (partially)
- Chem reaction tester (though this is debug, so it doesn't really matter)
- Portable chem mixer (partially)
- Mortar (partially)
- Syrup bottle
- Coffee cup
- Water bottle
- Sales Tagger
- Destructive Analyzer
- Tram remote assemblies
- Tram crossing signals

### LINKED ISSUE #82668: Unable to ventcrawl with alt+click
Reporting client version: 515.1635

## Issue Summary

Unable to crawl through vents with alt+click

## Round ID:
[227228](https://scrubby.melonmesa.com/round/227228)

## Testmerges:
- [Alt click refactor](https://www.github.com/tgstation/tgstation/pull/82656)

## Reproduction:

1. Be a rat King or a xeno near a vent.
2. Alt+click the vent to attempt to ventcrawl.
3. Observe you are unable to crawl into the vent.

### PR DESCRIPTION
**Pull request by jlsnow301**
## About The Pull Request
Rewrites how alt click works. 
Based heavily on #82625. What a cool concept, it flows nicely with #82533. 

Fixes #81242 
(tm bugs fixed)
Fixes #82668

<details><summary>More info for devs</summary>

Handy regex used for alt click s&r:
`AltClick\((.*).*\)(\n\t.*\.\.\(\))?`
`click_alt($1)` (yes I am aware this only copies the first arg. there are no other args!)

### Obj reskins
No reason for obj reskin to check on every single alt click for every object. It applies to only a few items.
- Moved to obj/item
- Made into signal
- Added screentips

### Ventcrawling
Every single atmospherics machine checked for ventcrawling capability on alt click despite only 3 objects needing that functionality. This has been moved down to those individual items.
</details>

## Why It's Good For The Game
For players: 
- Alt clicking should work more logically, not causing double actions like eject disk and open item window
- Added context menus for reskinnable items
- Removed adjacency restriction on loot panel

For devs:
- Makes alt click interactions easier to work with, no more click chain nonsense and redundant guard clauses.
- OOP hell reduced
- Pascal Case reduced
- Glorious snake case

## Changelog
:cl:
add: The lootpanel now works at range.
add: Screentips for reskinnable items.
fix: Alt click interactions have been refactored, which may lead to unintentional changes to gameplay. Report any issues, please.
/:cl:

### CODE REVIEW

File: code/_onclick/click_alt.dm [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * Override this if you want to change all alt-click behavior for a type.
+ */
+/atom/proc/base_click_alt(mob/user)
+	SHOULD_CALL_PARENT(TRUE)
+	PROTECTED_PROC(TRUE)
+
+	if(!can_interact(user))
+		return
+
+	var/early_sig_return = SEND_SIGNAL(src, COMSIG_CLICK_ALT, user) & COMPONENT_CANCEL_CLICK_ALT
```
Comment (MrMelbert) [inline code/_onclick/click_alt.dm:14]: IMO you should dump `COMPONENT_CANCEL_CLICK_ALT` and just let the signal return `CLICK_ACTION_SUCCESS` or `CLICK_ACTION_BLOCKING`

File: code/modules/projectiles/guns/ballistic.dm [hunk 2]
```diff
@@ -490,18 +490,21 @@
 		update_weight_class(w_class - I.w_class)
 	return ..()
 
-/obj/item/gun/ballistic/AltClick(mob/user)
+/obj/item/gun/ballistic/click_alt(mob/user)
 	if (unique_reskin && !current_skin && user.can_perform_action(src, NEED_DEXTERITY))
```
Comment (jlsnow301) [inline code/modules/projectiles/guns/ballistic.dm:494]: Remove

File: code/modules/projectiles/guns/ballistic.dm [hunk 3]
```diff
@@ -490,18 +490,21 @@
...
-	if(loc == user)
-		if(suppressed && can_unsuppress)
-			var/obj/item/suppressor/S = suppressed
-			if(!user.is_holding(src))
-				return ..()
-			balloon_alert(user, "[S.name] removed")
-			user.put_in_hands(S)
-			clear_suppressor()
+		return NONE
+	if(!loc == user)
```
Comment (jlsnow301) [inline code/modules/projectiles/guns/ballistic.dm:497]: Remove

File: code/modules/projectiles/guns/ballistic/shotgun.dm [hunk 4]
```diff
@@ -288,10 +289,10 @@
 	can_be_sawn_off = TRUE
 	pb_knockback = 3 // it's a super shotgun!
 
-/obj/item/gun/ballistic/shotgun/doublebarrel/AltClick(mob/user)
-	. = ..()
+/obj/item/gun/ballistic/shotgun/doublebarrel/click_alt(mob/user)
 	if(unique_reskin && !current_skin && user.can_perform_action(src, NEED_DEXTERITY))
```
Comment (jlsnow301) [inline code/modules/projectiles/guns/ballistic/shotgun.dm:293]: Remove

File: code/modules/reagents/reagent_containers/spray.dm [hunk 5]
```diff
@@ -438,9 +438,13 @@
...
-/obj/item/reagent_containers/spray/medical/AltClick(mob/user)
-	if(unique_reskin && !current_skin && user.can_perform_action(src, NEED_DEXTERITY))
-		reskin_obj(user)
+/obj/item/reagent_containers/spray/medical/click_alt(mob/user)
+	if(!user.can_perform_action(src, NEED_DEXTERITY))
+		return NONE
+	if(!unique_reskin || current_skin)
+		return CLICK_ACTION_BLOCKING
+	reskin_obj(user)
+	return CLICK_ACTION_SUCCESS
```
Comment (jlsnow301) [inline code/modules/reagents/reagent_containers/spray.dm:447]: Remove

File: code/game/objects/structures/lavaland/geyser.dm [hunk 6]
```diff
@@ -167,14 +167,12 @@
 
 	playsound(src, 'sound/machines/click.ogg', 10, TRUE)
 
-/obj/item/plunger/AltClick(mob/user)
-	if(!istype(user) || !user.can_perform_action(src))
-		return
-
+/obj/item/plunger/click_alt(mob/user)
 	var/new_layer = tgui_input_list(user, "Select a layer", "Layer", GLOB.plumbing_layers)
 	if(isnull(new_layer) || !user.can_perform_action(src))
```
Comment (jlsnow301) [inline code/game/objects/structures/lavaland/geyser.dm:172]: Reverse

**Review by jlsnow301 — COMMENTED** [review]
Notes bc I'm not home

**Comment by OverwatchVoice** [comment]
Alt clicking to eject IDs from PDAs and cigarettes from cigarette packs doesn't work.

**Comment by Metekillot** [comment]
wondrous, glorious, good

**Comment by TheRyeGuyWhoWillNowDie** [comment]
breaks like, every alt click interaction. no longer possible to check contents of storage item with alt click, no longer possible to empty fire extinguishers, etc. alt click menu also barely works and will always target the tile you're standing on which breaks being able to move items under flaps and whatnot

**Comment by 00-Steven** [comment]
having the current iteration of this pr open right now, you _can_ actually open storage with alt click and emptying fire extinguishers does work!
things I've tested so far:
- alt-clicking bags/boxes to open their contents
- turning on organ harvester works
- emptying fire extinguishers works
- adjusting (tailored) winter coats, labcoats, cargo gorkas, jumpsuits works
- adjusting bandanas works
- flipping caps works
- adding and removing pride pins/dogtags from jumpsuits works
- rotating sink and shower frames works
- rotating chairs, sofas, office chairs, stools works, inc while sitting on them
- rotating plumbing lines for hydroponics trays, smoke machines, and everything in the base/science/service plumbing constructors works
- maximizing pressure/heat/transfer rate on gas and volume pumps, passive and temperature gates, pressure valves, temperature pumps, and gas mixers and filters works
- rotating computer frames works
- rotating booze and soda dispensers works
- rotating wheelchairs works
- controller shell alternate signal works
- removing the cap from coffee cups and syrup bottles works
- reskinning medical sprays works
- sealing chem packs works
- racking dual tube shotguns works
- spinning the chamber on revolvers works
- removing arrows from bows works
- removing pens from clipboards works

**Comment by jlsnow301** [comment]
> breaks like, every alt click interaction. no longer possible to check contents of storage item with alt click, no longer possible to empty fire extinguishers, etc. alt click menu also barely works and will always target the tile you're standing on which breaks being able to move items under flaps and whatnot

This gets updated pretty quickly, so please make an issue report with a round ID

**Comment by jlsnow301** [comment]
> This is terrible currently. Can't even rotate shit with alt click. Did it override other settings or something?

dreamseeker_wwT4kyRhdo

**Comment by ophaq** [comment]
> > This is terrible currently. Can't even rotate shit with alt click. Did it override other settings or something?
> image


Yeah, I don't know why I couldn't rotate things. Maybe it was because I was a cyborg in the round? The new loot panel was pretty fucky when I was trying it out. It wouldn't display items on the ground in the box either when I tried alt clicking the floor.

**Comment by jlsnow301** [comment]
>  It wouldn't display items on the ground in the loot panel box either when I tried alt clicking the floor.

Could you elaborate, or better, write an issue with some repro steps?

**Comment by KingkumaArt** [comment]
Imagine you want to clear the pad after someone placed some orders on the express. before, you would walk up, alt click it, see the list, and click with your welder active on each listed drop pod. now, it looks like this and you can only click on one drop pod at a time.
https://i.imgur.com/14A9PQh.png

**Comment by jlsnow301** [comment]
wonder what this does
image

**Pull request merged**